### PR TITLE
Basis pow decomp

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
With `jax.jit` and traced input states, the decomposition of `BasisState` uses `RX` gates rather than the `PauliX` gates that can be used with `qjit` or static input states.
This is a problem even if we do not intend to use that branch because `BasisState` (or its decomp) are used in phase gradient decompositions, which then leads to the circular decomposition dependency `(some rotation) -> (BasisState) -> RX -> BasisState -> RX`...

**Description of the Change:**
Change the parametric branch to use `X(wires[i]) ** state[i]` instead of `RX` gates, getting rid of the problem (and decomposing to a usually cheaper gate).

**Benefits:**
Helps unblock phase gradient discretization decompositions.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/ops/qubit/state_preparation.py [hunk 1]
```diff
@@ -188,7 +188,12 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
 def _basis_state_decomp_resources(num_wires):
     # Represent one of the X gates as an RX and a GlobalPhase because RX is
     # used when jax-jit is enabled without capture/qjit.
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]:
```suggestion
```

File: pennylane/ops/qubit/state_preparation.py [hunk 2]
```diff
@@ -188,7 +188,12 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
...
-    return {qml.X: num_wires - 1 or num_wires, qml.RX: 1, qml.GlobalPhase: 1}
+    resources = {
+        pow_resource_rep(qml.X, base_params={}, z=0): 1,
+        pow_resource_rep(qml.X, base_params={}, z=1): 1,
+        qml.X: max(num_wires - 2, 1),
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: perhaps divide by 2 since that would be the expected average for random basis states?

File: pennylane/ops/qubit/state_preparation.py [hunk 3]
```diff
@@ -188,7 +188,12 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
 def _basis_state_decomp_resources(num_wires):
     # Represent one of the X gates as an RX and a GlobalPhase because RX is
     # used when jax-jit is enabled without capture/qjit.
-    return {qml.X: num_wires - 1 or num_wires, qml.RX: 1, qml.GlobalPhase: 1}
+    resources = {
+        pow_resource_rep(qml.X, base_params={}, z=0): 1,
+        pow_resource_rep(qml.X, base_params={}, z=1): 1,
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: what are those?

File: pennylane/ops/qubit/state_preparation.py [hunk 4]
```diff
@@ -209,11 +214,7 @@ def _basis_state_decomp(state, wires, **__):
         # This branch is for supporting jax-jit without capture/qjit. This is necessary if
         # the state is traced
         if abstract_state:
-            global_phase = 0.0
-            for wire, basis in zip(wires, state):
-                qml.RX(basis * np.pi, wires=wire)
-                global_phase += basis * np.pi / 2
-            qml.GlobalPhase(-global_phase)
+            _ = [qml.X(wires=wire) ** basis for wire, basis in zip(wires, state)]
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: very elegant and neat!
we could think long term about something that is more efficient like
```python
@qp.for_loop(0, len(wires)+1, 1)
def cond_xs(i):
    qp.cond(state[i], qml.X)(wires=wires[i])

cond_xs()
```

(I guess this wont work right now because the for loop problem persists?

**Review by Qottmann — APPROVED** [review]
thanks for the fix!

File: pennylane/ops/qubit/state_preparation.py (hunk 4 above)
Comment (Qottmann) [reply pennylane/ops/qubit/state_preparation.py]: for the controlled basis embedding that we use, we can also directly use this implementation with CNOTs instead of Xs

File: pennylane/ops/qubit/state_preparation.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: Yeah I think there is no unique meaning to those numbers for `exact=False`, so we can also do average cost 👍

File: pennylane/ops/qubit/state_preparation.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: Those are the power operators that appear in the new alternative decomposition (once with exponent `z=0` for the scenario where `state[i]=0` and once with `z=1`). We must report those here to trigger the graph system to account for them.

File: pennylane/ops/qubit/state_preparation.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: > we could think long term about something that is more efficient like

Isn't that what the other branch is doing? 🤔

Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: >for the controlled basis embedding that we use, we can also directly use this implementation with CNOTs instead of Xs

I was thinking that as well! Just added it to this branch  :)  (->re-requested review because it's a much bigger change now)

File: pennylane/ops/qubit/state_preparation.py [hunk 5]
```diff
@@ -186,9 +191,14 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
 
 
 def _basis_state_decomp_resources(num_wires):
-    # Represent one of the X gates as an RX and a GlobalPhase because RX is
-    # used when jax-jit is enabled without capture/qjit.
-    return {qml.X: num_wires - 1 or num_wires, qml.RX: 1, qml.GlobalPhase: 1}
+    # Represent one of the X gates as a power of X for z=0, 1 each.
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]:
```suggestion
    # Represent one of the X gates as X**z for z=0, 1 each.
```

X as a power of X sounds odd

File: pennylane/ops/qubit/state_preparation.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/ops/qubit/state_preparation.py]: as discussed offline, would be neat to separate the two decompositions depending on which branch we are on instead of covering both with one **IF** that is possible. Else looks fine as is 👍

File: pennylane/ops/qubit/state_preparation.py [hunk 6]
```diff
@@ -223,7 +229,55 @@ def _loop(i):
...
+
+
+def _single_ctrl_basis_state_decomp_resources(base_params, **_):
+    num_wires = base_params["num_wires"]
+    # Represent one of the CX gates as a power of CX for z=0, 1 each.
+    # used when jax-jit is used, without capture/qjit.
+    resources = {
+        pow_resource_rep(qml.CNOT, base_params={}, z=0): 1,
+        pow_resource_rep(qml.CNOT, base_params={}, z=1): 1,
+        qml.CNOT: max(num_wires - 2, 1),
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]:
```suggestion
        qml.CNOT: max((num_wires - 2) // 2, 1),
```

can also do the average case here

File: pennylane/ops/qubit/state_preparation.py [hunk 7]
```diff
@@ -223,7 +229,55 @@ def _loop(i):
...
+
+    @qml.for_loop(len(target_wires))
+    def _loop(i):
+        qml.cond(qml.math.allclose(state[i], 1), qml.CNOT)([cwire, target_wires[i]])
+
+    _loop()  # pylint: disable=no-value-for-parameter
+
+
 add_decomps(BasisState, _basis_state_decomp)
+add_decomps("C(BasisState)", _single_ctrl_basis_state_decomp)
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: more a curiosity question: why do we need to register the same decomp for two different ops? Or rather, why do we register a decomp for `BasisState`, thats not an op no?

Comment (Qottmann) [reply pennylane/ops/qubit/state_preparation.py]: Like is that even necessary here? If not, we should move the decomps closer to the OP it actually decomposes (but I might be totally off here)

File: tests/ops/qubit/test_state_prep.py [hunk 8]
```diff
@@ -45,10 +45,12 @@ def test_assert_valid(self):
         qml.ops.functions.assert_valid(op, skip_differentiation=True)
 
     @staticmethod
-    def make_abstract_check(state_traced, wires_traced, closure_state, closure_wires):
+    def make_abstract_check(state_traced, wires_traced, closure_state, closure_wires, ctrld):
         """Create a function for JIT-related tests that creates tapes from the decomposition
         rules registered for `BasisState` with the graph-based decomposition system."""
 
+        op_name = "C(BasisState)" if ctrld else "BasisState"
```
Comment (Qottmann) [inline tests/ops/qubit/test_state_prep.py]: It's neat this way because you dont need to add an extra test
This is very much up to taste, for me personally I'd prefer a separate test that is easier to follow, rather than making this test more convoluted and checking multiple cases 😅

File: pennylane/ops/qubit/state_preparation.py (hunk 7 above)
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: Not sure about the conventions, but since we specifically only decompose the case for one control, does it have to be `1C(BasisState)`? Is that even a thing? Sorry for the perhaps very naive question 😅

**Review by Qottmann — APPROVED** [review]
No major blockers, looks good overall!

Would be nice to separate the decomps for catpure enabled and not

The thing about testing is very subjective and up to personal preference, so I leave this at your discretion

File: pennylane/ops/qubit/state_preparation.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/ops/qubit/state_preparation.py]: after more discussion, doesnt seem feasible because we'd need info on state which is not available at compile time

File: pennylane/ops/qubit/state_preparation.py [hunk 9]
```diff
@@ -186,9 +191,14 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
...
 
 def _basis_state_decomp_resources(num_wires):
-    # Represent one of the X gates as an RX and a GlobalPhase because RX is
-    # used when jax-jit is enabled without capture/qjit.
-    return {qml.X: num_wires - 1 or num_wires, qml.RX: 1, qml.GlobalPhase: 1}
+    # Represent one of the X gates as a power of X for z=0, 1 each.
+    # used when jax-jit is used, without capture/qjit.
+    resources = {
+        pow_resource_rep(qml.X, base_params={}, z=0): 1,
+        pow_resource_rep(qml.X, base_params={}, z=1): 1,
```
Comment (Qottmann) [inline pennylane/ops/qubit/state_preparation.py]: should we still give those a more realistic estimate? I.e. something like this?

```suggestion
        pow_resource_rep(qml.X, base_params={}, z=0): max((num_wires - 2) // 2,
        pow_resource_rep(qml.X, base_params={}, z=1): max((num_wires - 2) // 2,
```

File: pennylane/ops/qubit/state_preparation.py [hunk 10]
```diff
@@ -223,7 +229,55 @@ def _loop(i):
     _loop()  # pylint: disable=no-value-for-parameter
 
 
+def _single_ctrl_basis_state_decomp_cond(num_control_wires, num_zero_control_values, **_):
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py]: Why do we need to manually define a control decomposition?

Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: Because otherwise the only path accessible to the graph system is via `C(uncontrolled_decomposition)`, which leads to `C(Pow(X))` entering the graph, which in turn pulls in a large number of expensive decompositions.

Yesterday I found out that this PR will not solve our problems entirely, so I'll modify it a bit.

File: pennylane/ops/qubit/state_preparation.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: The source for this is that we have two Operators that are virtually the same, but we did not make them simple aliases of each other, for some reason.

File: tests/ops/qubit/test_state_prep.py (hunk 8 above)
Comment (dwierichs) [reply tests/ops/qubit/test_state_prep.py]: Removed the dedicated controlled decomposition, so can remove the test 👍

File: pennylane/ops/qubit/state_preparation.py (hunk 7 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: multi-controlled things are printed by specs as `nC(...)` but I think not in the decomp system.
The fact that only one control wire can be handled is communicated via the registered condition function, so should be ok :)

File: pennylane/ops/qubit/state_preparation.py (hunk 10 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: Removed ✅

File: pennylane/ops/qubit/state_preparation.py [hunk 11]
```diff
@@ -185,45 +190,65 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
         return math.convert_like(ket, prep_vals)
 
 
+def _fallback_branch_resources(num_wires):
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py]: Maybe instead of "fallback" we just say "jit_compatible" or something?

File: pennylane/ops/qubit/state_preparation.py [hunk 12]
```diff
@@ -185,45 +190,65 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
...
+
+
+@register_condition(_fallback_branch_cond)
+@register_resources(_fallback_branch_resources, exact=False)
+def _fallback_branch_decomp(state, wires, **__):
+    if qml.math.is_abstract(state):
+        _ = [qml.X(wires=wire) ** basis for wire, basis in zip(wires, state)]
+        return
+    _ = [qml.X(wires=wire) for wire, basis in zip(wires, state) if basis]
+
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:219]: Should we have this in the rule at all? If this was possible, it would take the `_basis_state_decomp` pathway.

Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py:219]: We could leave it out. Note, though, that the check whether we are using `jax.jit` can lead to false positives for the question `is_abstract(state)?` 
So this branch is actually being hit when using `jax.jit` with a non-traced state

Comment (albi3ro) [reply pennylane/ops/qubit/state_preparation.py:219]: I think I would be fine with just having the abstract branch.  Pow(X) isn't really that problematic, and will just decompose to the non-abstract branch if that's ever needed.

Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py:219]: ✅

File: tests/ops/qubit/test_state_prep.py (hunk 8 above)
Comment (Qottmann) [reply tests/ops/qubit/test_state_prep.py]: win win 🚀

**Review by Qottmann — APPROVED** [review]
wow, super clean now!

File: pennylane/ops/qubit/state_preparation.py (hunk 7 above)
Comment (Qottmann) [reply pennylane/ops/qubit/state_preparation.py]: :notlikethis: 
but would consider this out of scope for this fix 👍

File: pennylane/ops/qubit/state_preparation.py [hunk 13]
```diff
@@ -185,45 +190,61 @@ def state_vector(self, wire_order: WiresLike | None = None) -> TensorLike:
...
+def _jax_jit_basis_state_cond(**_):
+    if qml.capture.enabled() or qml.compiler.active():
+        return False
+
+    try:
+        import jax  # pylint: disable=import-outside-toplevel,unused-import
+    except ModuleNotFoundError:
+        # Can't be using jax.jit if jax is not installed.
+        return False
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py]: potentially a good use for `importlib.utils.find_spec` here. Been encouraged to start using that instead of try-except blocks for imports.

File: tests/ops/qubit/test_state_prep.py [hunk 14]
```diff
@@ -60,7 +60,9 @@ def abstract_check(state, wires):
             assert qml.math.is_abstract(state) == state_traced
             assert qml.math.is_abstract(wires) == wires_traced
             tapes = []
-            for rule in qml.list_decomps(qml.BasisState):
+            for rule in qml.list_decomps("BasisState"):
```
Comment (albi3ro) [inline tests/ops/qubit/test_state_prep.py]: Why the change here? Is there something wrong with doing `qml.list_decomps(qml.BasisState)`?

File: pennylane/ops/qubit/state_preparation.py (hunk 13 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: ah nice, I'll look into that!

File: tests/ops/qubit/test_state_prep.py (hunk 14 above)
Comment (dwierichs) [reply tests/ops/qubit/test_state_prep.py]: Ah, this was because I used to have the alternative `"C(BasisState)"` here as well... Will revert.

File: pennylane/ops/qubit/state_preparation.py (hunk 13 above)
Comment (dwierichs) [reply pennylane/ops/qubit/state_preparation.py]: Done ✅

File: tests/ops/qubit/test_state_prep.py (hunk 14 above)
Comment (dwierichs) [reply tests/ops/qubit/test_state_prep.py]: Done ✅

File: pennylane/templates/embeddings/basis.py [hunk 15]
```diff
@@ -16,7 +16,10 @@
 """
 
 from pennylane.decomposition import add_decomps
-from pennylane.ops.qubit.state_preparation import BasisState, _basis_state_decomp
+from pennylane.ops.qubit.state_preparation import (
+    BasisState,
+    _basis_state_decomp,
+)
```
Comment (astralcai) [inline pennylane/templates/embeddings/basis.py]: What changed here?

Comment (dwierichs) [reply pennylane/templates/embeddings/basis.py]: I originally added more imports here for something else, which was deleted again.
The left-over comma prevents black (_still_ 😩 ) from simplifying the formatting here :/ Reverting...

Comment (dwierichs) [inline pennylane/templates/embeddings/basis.py]:
```suggestion
from pennylane.ops.qubit.state_preparation import BasisState, _basis_state_decomp
```

**Review by astralcai — APPROVED** [review]
Thanks for finding a solution for this!
